File: benchmarks/server/driver/src/drivers/sirannon.ts
import type { FailureClassifier, FailureKind } from '../failures.ts'
import type { RemoteDatabase } from '../sirannon-client.ts'
import { SirannonClient } from '../sirannon-client.ts'
import type { SeedTable } from '../workloads/workload.ts'
import { Driver, type TransactionStatement } from './driver.ts'

const SEED_BATCH_ROWS = 25_000

const SIRANNON_SHED = new Set(['WRITE_OVERLOADED'])
const SIRANNON_TIMEOUT = new Set(['TIMEOUT', 'WRITER_WORKER_TIMEOUT'])
const SIRANNON_CONNECTION = new Set(['CONNECTION_ERROR', 'TRANSPORT_ERROR'])
const SIRANNON_CLIENT = new Set(['INVALID_ARGUMENT', 'INVALID_RESPONSE', 'ROUTING_ERROR'])
const SIRANNON_UNKNOWN = new Set(['UNKNOWN_ERROR'])

function remoteCodeOf(err: unknown): string | undefined {
  const code = (err as { code?: unknown }).code
  return typeof code === 'string' && code !== '' ? code : undefined
}

function sirannonKind(code: string, err: unknown): FailureKind {
  // Re-reads err because sirannonCode substitutes err.name when the SDK reported no code at all.
  if (remoteCodeOf(err) === undefined) {
    return 'client_error'
  }
  if (SIRANNON_SHED.has(code)) {
    return 'shed'
  }
  if (SIRANNON_TIMEOUT.has(code)) {
    return 'timeout'
  }
  if (SIRANNON_CONNECTION.has(code)) {
    return 'connection'
  }
  if (SIRANNON_CLIENT.has(code)) {
    return 'client_error'
  }
  if (SIRANNON_UNKNOWN.has(code)) {
    return 'unknown'
  }
  return 'server_error'
}

function sirannonCode(err: unknown): string {
  return remoteCodeOf(err) ?? (err instanceof Error ? err.name : typeof err)
}

export class SirannonDriver extends Driver {
  readonly name = 'sirannon'
  readonly delivery = 'websocket'
  readonly dialect = 'sqlite'

  private readonly baseUrl: string
  private readonly databaseId: string
  private readonly endpoint: string
  private readonly durability: string
  private readonly requestTimeoutMs: number
  private client: SirannonClient | null = null
  private db: RemoteDatabase | null = null
  readonly failureClassifier: FailureClassifier = { codeOf: sirannonCode, kindOf: sirannonKind }

  constructor(baseUrl: string, databaseId: string, durability: string, requestTimeoutMs: number) {
    super()
    this.baseUrl = baseUrl
    this.databaseId = databaseId
    this.endpoint = `${baseUrl.replace(/\/+$/, '')}/db/${encodeURIComponent(databaseId)}`
    this.durability = durability
    this.requestTimeoutMs = requestTimeoutMs
  }

  async connect(): Promise<void> {
    this.client = new SirannonClient(this.baseUrl, { requestTimeout: this.requestTimeoutMs })
    this.db = this.client.database(this.databaseId)
    await this.read('SELECT 1', [])
  }

  private database(): RemoteDatabase {
    if (this.db === null) {
      throw new Error('Sirannon driver used before connect()')
    }
    return this.db
  }

  private async post(path: string, body: unknown): Promise<Record<string, unknown>> {
    const response = await fetch(`${this.endpoint}${path}`, {
      method: 'POST',
      headers: { 'content-type': 'application/json' },
      body: JSON.stringify(body),
    })
    if (!response.ok) {
      throw new Error(`POST ${path} returned ${response.status}: ${await response.text()}`)
    }
    return (await response.json()) as Record<string, unknown>
  }

  async info(): Promise<Record<string, unknown>> {
    // A PRAGMA synchronous read would report the read pool's NORMAL, not the writer's durability.
    const settings: Record<string, unknown> = {
      engine: 'sirannon',
      delivery: this.delivery,
      durability_requested: this.durability,
    }
    const journalRows = (await this.database().query('PRAGMA journal_mode')) as Array<Record<string, unknown>>
    const firstJournal = journalRows[0]
    settings.journal_mode = firstJournal ? String(Object.values(firstJournal)[0] ?? 'unknown') : 'unknown'
    const versionRows = (await this.database().query('SELECT sqlite_version() AS version')) as Array<
      Record<string, unknown>
    >
    settings.version = versionRows[0] ? String(versionRows[0].version) : 'unknown'
    return settings
  }

  async executeDdl(statements: string[]): Promise<void> {
    for (const statement of statements) {
      await this.post('/execute', { sql: statement })
    }
  }

  async dropTables(tables: string[]): Promise<void> {
    for (const table of tables) {
      await this.post('/execute', { sql: `DROP TABLE IF EXISTS ${table}` })
    }
  }

  async seed(tables: SeedTable[]): Promise<void> {
    for (const table of tables) {
      await this.database().loadAll(this.insertSql(table), table.rows, {
        batchSize: SEED_BATCH_ROWS,
        durability: 'off',
      })
    }
  }

  async read(sql: string, params: unknown[]): Promise<void> {
    await this.database().query(sql, params)
  }

  async write(sql: string, params: unknown[]): Promise<void> {
    await this.database().execute(sql, params)
  }

  async transaction(statements: TransactionStatement[]): Promise<void> {
    await this.database().transaction(statements)
  }

  transactionRoundTrips(_statementCount: number): number {
    return 1
  }

  async close(): Promise<void> {
    if (this.client !== null) {
      this.client.close()
      this.client = null
      this.db = null
    }
  }
}

File: benchmarks/server/src/sirannon_bench/drivers/sirannon.py
"""Drive Sirannon through its real HTTP server front-end.

This reproduces exactly what the SirannonClient SDK sends: JSON bodies to
``POST /db/<id>/query``, ``/execute``, and ``/transaction`` with ``?`` placeholders and
positional parameters, no authentication, over a keep-alive connection pool. Charging Sirannon
its real HTTP and JSON cost is the point; that cost is its shipping client path.
"""

from __future__ import annotations

import httpx

from ..core.driver import Driver
from ..core.workloads import SeedTable

_SEED_CHUNK = 500


class SirannonHttpDriver(Driver):
    name = "sirannon"
    delivery = "http"
    dialect = "sqlite"

    def __init__(self, base_url: str, database_id: str, durability: str, max_in_flight: int) -> None:
        self._endpoint = f"{base_url.rstrip('/')}/db/{database_id}"
        self._durability = durability
        self._max_in_flight = max_in_flight
        self._client: httpx.AsyncClient | None = None

    async def connect(self) -> None:
        limits = httpx.Limits(
            max_connections=self._max_in_flight,
            max_keepalive_connections=self._max_in_flight,
        )
        self._client = httpx.AsyncClient(timeout=httpx.Timeout(30.0), limits=limits)
        await self.read("SELECT 1", [])

    def _http(self) -> httpx.AsyncClient:
        if self._client is None:
            raise RuntimeError("Sirannon driver used before connect()")
        return self._client

    async def _post(self, path: str, body: dict) -> dict:
        response = await self._http().post(f"{self._endpoint}{path}", json=body)
        if response.status_code < 200 or response.status_code >= 300:
            raise RuntimeError(f"POST {path} returned {response.status_code}: {response.text}")
        return response.json()

    async def info(self) -> dict:
        # The read pool always opens at synchronous=NORMAL, so a PRAGMA read over HTTP would
        # misreport the writer's durability. The authoritative durability is the requested value,
        # which the server applies to the writer connection; journal_mode is persistent per
        # database, so it reads back accurately from any connection.
        settings: dict[str, str] = {"engine": "sirannon", "delivery": "http", "durability_requested": self._durability}
        journal_rows = (await self._post("/query", {"sql": "PRAGMA journal_mode"})).get("rows") or []
        first = journal_rows[0] if journal_rows else {}
        settings["journal_mode"] = str(next(iter(first.values()), "unknown")) if first else "unknown"
        version_rows = (await self._post("/query", {"sql": "SELECT sqlite_version() AS version"})).get("rows") or []
        settings["version"] = str(version_rows[0].get("version")) if version_rows else "unknown"
        return settings

    async def execute_ddl(self, statements: list[str]) -> None:
        for statement in statements:
            await self._post("/execute", {"sql": statement})

    async def drop_tables(self, tables: list[str]) -> None:
        for table in tables:
            await self._post("/execute", {"sql": f"DROP TABLE IF EXISTS {table}"})

    async def seed(self, tables: list[SeedTable]) -> None:
        for table in tables:
            if not table.rows:
                continue
            insert = self.insert_sql(table)
            for offset in range(0, len(table.rows), _SEED_CHUNK):
                chunk = table.rows[offset : offset + _SEED_CHUNK]
                statements = [{"sql": insert, "params": row} for row in chunk]
                await self._post("/transaction", {"statements": statements})

    async def read(self, sql: str, params: list) -> None:
        await self._post("/query", {"sql": sql, "params": params})

    async def write(self, sql: str, params: list) -> None:
        await self._post("/execute", {"sql": sql, "params": params})

    async def close(self) -> None:
        if self._client is not None:
            await self._client.aclose()
            self._client = None
